CMake: Generate pkg-config files for all Windows builds

fanc999 opened the pull request:
Hi,

This PR is opened to support generating pkg-config files also for native (non-Cygwin) Windows builds, in addition to MinGW/mingw-w64 builds, since the Meson build system (which uses pkg-config and CMake to find dependencies for most packages) is becoming more widespread for building items such as Cairo and the GNOME/GTK stack and the current CMake config files generated by the CMake builds are not usable by Meson as a consequence of #179, which should be fixed in a PR of its own.

With blessings, thank you!

fanc999-1 committed:
CMake: Generate pkg-config files for all Windows builds

The Meson build system makes use of pkg-config files or CMake to find most of
the dependencies via its "built-in" dependency search system.  As libpng is a
very widely used package, this makes the CMake build system generate pkg-config
files for libpng even for non-Cygwin/non-MinGW Windows builds, so that Meson
could locate libpng via pkg-config as well even on Visual Studio builds, for
instance.

We also do the following on native, non-MinGW Windows builds:

*  Replace -lpngXX with libpngXX.lib, for the libpng library that we link to
*  Use the correct ZLib library .lib name for `libs.private`.
*  Replace `Requires.private: zlib` with `Requires.private:`, since we already
   say in `libs.private` that we are linking to ZLib

ctruta commented:
Thank you, looks neat, but could you please fix the following build error? I tested it on macOS, and it fails with the Ninja generator and the Release target, as follows:
```
CMake Error at cmake_install.cmake:90 (file):
  file INSTALL cannot find
  "/Users/ctruta/src/libpng/build/libpng16-config": No such file
  or directory.

FAILED: CMakeFiles/install.util
```

BotellaA commented:
Any update on this? It will be very helpful!

fanc999 commented:
Hi @ctruta,

Sorry for the delay, as I was away and this PR was out for some time... I rebased some items, so I think things should probably be remedied, hopefully, since I don't have a macOS devenv.

ctruta commented on CMakeLists.txt:
```diff
@@ -931,30 +931,42 @@ endif()
+else()
+  STRING (TOLOWER "${CMAKE_BUILD_TYPE}" config_lower)
+  if (config_lower STREQUAL "debug")
+    set(LIBS    "${ZLIB_LIBRARY_DEBUG} ${M_LIBRARY}")
```
I'm not an expert at how CMake libraries are being exported and imported. My understanding from other people is that there shouldn't be a `ZLIB_LIBRARY` or a `ZLIB_INCLUDE_DIR` (or a `ZLIB_LIBRARY_DEBUG`, etc.), but rather a symbol named `ZLIB_LIBRARIES` and another symbol named `ZLIB_INCLUDE_DIRS`.

For that reason, I am asking: are you sure that we shouldn't just have the following:
```
  set(LIBS "${ZLIB_LIBRARIES} ${M_LIBRARY}"
```

I can imagine that there are pros and cons to each side, but I would like to know your opinion also.

Moreover, take a look at line 598, further up in this CMake file:
```
  target_link_libraries(png ${ZLIB_LIBRARIES} ${M_LIBRARY})
```

I'd like to be sure that there won't be breakages due to this inconsistency.

fanc999 replied:
Hi,

Sadly, using `${ZLIB_LIBRARIES}` in the pkg-config file generating won't work, because you will have all the variants inside there.  Will look for a cleaner solution.

fanc999 replied:
Hi,

I updated the CMakeLists.txt to reflect on the suggestions that were raised purely on the part that this PR attempts to cover, so I did not replace the occurences of `ZLIB_LIBRARIES` with `ZLIB::ZLIB` in any way but instead grabbed the library names for our pkg-config files with `get_target_property(ZLIB_LIB ZLIB::ZLIB IMPORTED_LOCATION[_xxx])`.

ctruta commented on CMakeLists.txt:
```diff
@@ -931,30 +931,42 @@ endif()
     CLEAN_DIRECT_OUTPUT 1)
+  if(WIN32 AND NOT CYGWIN AND NOT MINGW)
+    foreach (pc_file ${CMAKE_CURRENT_BINARY_DIR}/${PNGLIB_NAME}.pc ${CMAKE_CURRENT_BINARY_DIR}/libpng.pc)
+      FILE(READ ${pc_file} png_pc_orig)
```
Could you please use lowercase CMake commands (`file`, `string`, etc.), for the sake of stylistic consistency?

ctruta reviewed (changes requested):
Overall comment: this pull request appears to be carefully developed, and I like it a lot. There are just a few comments to be addressed.

fanc999-1 committed:
CMakeLists.txt: Fix stylelistic issues

Use all-lowercase for CMake commands

fanc999-1 committed:
CMakeLists.txt: Improve pkg-config generating on Windows

Use the target properties of ZLIB::ZLIB to get the library names for the
pkg-config files that we generate, and unify things with other builds.
We need to use the IMPORT_LOCATION[_$<CONFIG>] for ZLib since it does not
define IMPORT_IMPLIB[_$<CONFIG>] even on Windows DLL builds, and assume
that the $<CONFIG> we use above is RELEASE for both MinSizeRel and
RelWithDebInfo builds.

ctruta commented:
Hi, here's a heads-up,

I am currently applying refactoring changes in order to clean up the CMakefile and to make Clang on Windows work correctly, without regressing other platforms. Your contributions from this pull request may or may not break because of these upcoming changes.

On the positive side of the news, this pull request is next on my to-do list.
